WIP: Delivery workflow

johandahlberg opened the pull request:
This implements the delivery workflow for a runfolder for the following stages so far:

 - [x] Check which projects are available for runfolder
 - [x] Get emails for PIs for these projects
 - [x] Get PI id for PIs in SUPr
 - [x] Stage projects
 - [x] Create delivery projects for PIs
 - [x] Run and monitor delivery with mover

It should be ready for some initial code review at this point, but it's not yet ready to be merged.

In addition to this we will need to implement running deliveries on arbitrary folder structures - so that e.g. WGS projects can also be delivered.

johanherman commented on actions/arteria_delivery_service.py:
```diff
@@ -0,0 +1,68 @@
+
+        links_results = dict(map(lambda x: (x, None), links))
+
+        while not all(map(is_in_end_state, links_results.values())):
```
Perhaps you should watch out for unexpected responses (states) and/or errors from the server side around here. If I'm reading this correctly this could end up in an infinite loop otherwise.

johandahlberg replied:
Good point. Will look at it. It will be canceled eventually by the action timeout, however.

johanherman commented on actions/supr.py:
```diff
@@ -0,0 +1,91 @@
+            # realistic in here later.
+            # 'allocated': size_of_delivery,
+            'api_opaque_data': '',
+            'ngi_ready': False,
```
When do you think we would like to set this to `True`?

johandahlberg replied:
From the docs:
`* ngi_ready: Flag (boolean) to tell if delivery of data from NGI has finished. Only used for NGI Delivery projects.`
I think probably this is a flag that Mover should set in the end...

johanherman replied:
When we're calling Mover directly as a shell command then yes, one could argue that Mover can set this itself. But if we want Mover to poll SUPR for ready projects to deliver then we should set this to `True` whenever we want to delegate the responsibility. So perhaps we could set it to `True` just before we call Mover in the mean time?

johandahlberg replied:
As far as I understand that is not how things are going to work, at least short term Mover will not poll Supr for info about projects.

b97pla replied:
I've been a bit confused about this flag but I think @johanherman has it right, i.e. this is a flag that NGI will use to indicate for Uppmax that a project is ready to be delivered. This is what it was initially defined as. Along the way I got the impression that it would be a flag mover would use to signal whether the transfer to Grus is complete but I realize that's probably not correct. Do you agree @johanherman?

johandahlberg replied:
Ok. So then we should set this to true once the delivery has finished successfully?

johandahlberg replied:
Not after? Since the docs say:

`ngi_ready: Flag (boolean) to tell if delivery of data from NGI has finished. Only used for NGI Delivery projects.`

I interpret this as something that should be true when the delivery is finished... And the delivery is not finished until Mover is done?

johanherman replied:
I had to check my mails. This (and other connected flags) have been going back and forth. And I guess you're also right @johandahlberg -- latest mail from Ankerstål on the 11th of Oct that I could find then he also references `ngi_ready` as it should be set by UPPMAX when Mover is finished, for the *user* to see what he can import data from. 

But we've also had these fields up for discussion from our perspective before, so, I don't know... perhaps it should be like this for now then..

b97pla replied:
@johanherman this is my impression as well, we should clear this up with Peter. Sounds good to leave it until we know for sure.

johandahlberg commented on actions/workflows/delivery_runfolder_workflow.yaml:
```diff
@@ -0,0 +1,96 @@
+                supr_api_user: <% task(get_config).result.result.supr_api_user %>
+                supr_api_key: <% task(get_config).result.result.supr_api_key %>
+                supr_api_url: <% task(get_config).result.result.supr_api_url %>
+                supr_api_url: <% task(get_config).result.result.supr_api_url %>
```
:+1:

johandahlberg commented on actions/workflows/delivery_runfolder_workflow.yaml:
```diff
@@ -0,0 +1,96 @@
+            projects_on_runfolder:
+              action: core.http
+              input:
+                url: "<% $.delivery_service_url %>/api/1.0/runfolders/<% $.runfolder_name %>/projects"
```
I think it's better to keep it here, since we might want to have multiple versions of the api in the future. However, it's of course a matter of taste exactly where to specify it.

johandahlberg commented on actions/workflows/delivery_runfolder_workflow.yaml:
```diff
@@ -0,0 +1,96 @@
+              input:
+                file_path: <% $.projects_pi_email_file %>
+                projects:
+                  projects: <% $.projects_on_runfolder %>
```
Because as far as I was able to see `list` is not a valid input type, but `object` is. Thus this somewhat contrived way of doing things.

johanherman replied:
Aha. I'm lövin it. :+1:

johandahlberg commented on actions/read_projects_email_file.py:
```diff
@@ -0,0 +1,30 @@
+            self.logger.info("Projects given and projects found in file did match...")
+            return True, result
+        else:
+            self.logger.info("Projects given and projects found in file did not match!")
```
:+1:

johandahlberg commented on actions/arteria_delivery_service.py:
```diff
@@ -0,0 +1,68 @@
+        if projects:
+            payload = {'projects': projects['projects']}
+        else:
+            payload = {}
```
Actually `projects_on_runfolder` will return all projects that are on the runfolder, so this might be a bit misleading. It this workflow having no projects at all specified shouldn't really happen...

johandahlberg commented:
I've addressed some the issues you raised in your code review @johanherman. I still think we should wait with merging this until we have the full workflow down, though.

johanherman commented on actions/arteria_delivery_service.py:
```diff
@@ -14,15 +14,20 @@ class ArteriaDeliveryService(Action):
                 if state == STAGING_SUCCESSFUL or state == STAGING_FAILED:
                     continue
-                else:
+                elif state in valid_states:
```
You don't think we should also watch out for e.g. `KeyError`, `ValueError` and `RequestException` or similar?

johandahlberg replied:
Well, if any of those are raised it will kill the execution right? So while the error message might not be super nice it will still give the result we're hoping for of killing the execution?

johanherman replied:
Sure. I was just thinking it is nice with elementary housekeeping stuff. Might make it easier to read in the Stackstorm output e.g.

johanherman commented:
Sure, we'll wait. :+1:

johandahlberg commented:
Perhaps you want to review this now @johanherman and then maybe we should merge it. I'm working on setting up new stories for the coming sprints, which are more specific than the old once relating to this.

johanherman commented on actions/arteria_delivery_service.py:
```diff
@@ -45,29 +51,125 @@ def is_in_end_state(link_state):
+                exit_status = False
+                self.logger.error('Project: {} was not successfully staged. '
+                                  'Had status: {}. Link was: '.format(project, status, link))
+            result[project] = int(link.split('/')[-1])
```
What is this magic?

johanherman commented on actions/arteria_delivery_service.py:
```diff
@@ -45,29 +51,125 @@ def is_in_end_state(link_state):
+
+class ArteriaStagingQuerier(ArteriaQuerierBase):
+
+    staging_successful = 'staging_successful'
```
Although it is pretty with the inheritance -- is there some specific reason for wanting to have "type specific" statuses? Instead of having, lets say, a general "failed", "pending", "in progress", etc. I guess the caller knows when he's speaking to the staging part and the delivery part.

johandahlberg commented:
This should be ready to merge now @johanherman.
